cdap-ui/app/cdap/components/DataPrep/Directives/MapToTarget/index.js

```diff
@@ -14,36 +14,367 @@
  * the License.
  */
 
-import React, { Component } from 'react';
+import React, { useEffect, useState } from 'react';
 import PropTypes from 'prop-types';
 import T from 'i18n-react';
+import { makeStyles } from '@material-ui/core';
+import Input from '@material-ui/core/Input';
+import LinearProgress from '@material-ui/core/LinearProgress';
+import List from '@material-ui/core/List';
+import ListItem from '@material-ui/core/ListItem';
+import ListItemText from '@material-ui/core/ListItemText';
+import { UncontrolledTooltip } from 'reactstrap';
 import classnames from 'classnames';
+import { preventPropagation, connectWithStore } from 'services/helpers';
+import { setPopoverOffset } from 'components/DataPrep/helper';
+import DataPrepStore from 'components/DataPrep/store';
+import {
+  execute,
+  setError,
+  loadTargetDataModelStates,
+  saveTargetDataModelStates,
+  setTargetDataModel,
+  setTargetModel
+} from 'components/DataPrep/store/DataPrepActionCreator';
+
+const useStyles = makeStyles(theme => ({
+  secondLevelPopover: {
+    width: '300px !important',
+  },
+  selectedItem: {
+    display: 'flex',
+    flexDirection: 'row',
+  },
+  selectedItemLabel: {
+    whiteSpace: 'nowrap',
+    textOverflow: 'ellipsis',
+    overflow: 'hidden',
+    fontWeight: 'bold',
+  },
+  selectedItemName: {
+    flex: 1,
+    whiteSpace: 'nowrap',
+    textOverflow: 'ellipsis',
+    overflow: 'hidden',
+  },
+  unselectIcon: {
+    cursor: 'pointer',
+    padding: theme.spacing(0.75),
+    margin: '0 !important',
+    '&:hover': {
+      fontWeight: 'bold',
+    },
+  },
+  optionSearch: {
+    width: '100%',
+    marginBottom: theme.spacing(0.5),
+  },
+  targetOptionList: {
+    overflowX: 'hidden',
+    overflowY: 'auto',
+    maxHeight: '400px',
+  },
+  targetOption: {
+    whiteSpace: 'nowrap',
+    textOverflow: 'ellipsis',
+    overflow: 'hidden',
+  },
+  highlight: {
+    color: theme.palette.primary.contrastText,
+    backgroundColor: theme.palette.primary.dark,
+  },
+}));
 
 const PREFIX = 'features.DataPrep.Directives.MapToTarget';
 
-export default class MapToTarget extends Component {
-  static propTypes = {
-    isOpen: PropTypes.bool,
-    isDisabled: PropTypes.bool,
-    column: PropTypes.string,
-    onComplete: PropTypes.func,
+const MapToTarget = (props) => {
+  const classes = useStyles(undefined);
+  const {
+    isOpen,
+    isDisabled,
+    column,
+    onComplete,
+    close,
+    dataModelList,
+    targetDataModel,
+    targetModel,
+  } = props;
+  const [loading, setLoading] = useState('');
+  const [searchText, setSearchText] = useState('');
+
+  useEffect(() => {
+    let pending = true;
+    setLoading(`${PREFIX}.initializingText`);
+    (async () => {
```

Comment:
Please use a named function here. 

```
useEffect(() => {
  ...
  async function loadTargetModel() {
    try {
        await loadTargetDataModelStates();
    } catch (error) {
        setError(error);
    } finally {
        if (pending) {
          setLoading('');
        }
    }
  }
  loadTargetModel();
});
```
Makes it slightly better for readability. Better yet if you could extract function outside of useEffect.

Reply:
Refactored to named function. I can extract this function outside of `useEffect` but it requires to move `pending` flag outside as well. The purpose of the `pending` flag is to prevent `setLoading` call on possibly detached component.